File: educational-core/src/com/jetbrains/edu/learning/checker/TaskCheckerProvider.java
package com.jetbrains.edu.learning.checker;

import com.intellij.openapi.project.Project;
import com.jetbrains.edu.learning.codeforces.checker.CodeforcesTaskChecker;
import com.jetbrains.edu.learning.codeforces.checker.CodeforcesTaskWithFileIOTaskChecker;
import com.jetbrains.edu.learning.codeforces.courseFormat.CodeforcesTask;
import com.jetbrains.edu.learning.codeforces.courseFormat.CodeforcesTaskWithFileIO;
import com.jetbrains.edu.learning.courseFormat.tasks.*;
import com.jetbrains.edu.learning.courseFormat.tasks.choice.ChoiceTask;
import com.jetbrains.edu.learning.handlers.CodeExecutor;
import com.jetbrains.edu.learning.handlers.DefaultCodeExecutor;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public interface TaskCheckerProvider {
    @NotNull
    TaskChecker<EduTask> getEduTaskChecker(@NotNull EduTask task, @NotNull Project project);

    @NotNull
    default OutputTaskChecker getOutputTaskChecker(@NotNull OutputTask task, @NotNull Project project, @NotNull CodeExecutor codeExecutor) {
        return new OutputTaskChecker(task, project, codeExecutor);
    }

    @NotNull
    default TheoryTaskChecker getTheoryTaskChecker(@NotNull TheoryTask task, @NotNull Project project) {
        return new TheoryTaskChecker(task, project);
    }

    @Nullable
    default TaskChecker<ChoiceTask> getChoiceTaskChecker(@NotNull ChoiceTask task, @NotNull Project project) {
        return task.getCanCheckLocally() ? new ChoiceTaskChecker(task, project) : null;
    }

    @Nullable
    default TaskChecker<CodeTask> getCodeTaskChecker(@NotNull CodeTask task, @NotNull Project project) {
        return null;
    }

    @NotNull
    default TaskChecker<IdeTask> getIdeTaskChecker(@NotNull IdeTask task, @NotNull Project project) {
        return new IdeTaskChecker(task, project);
    }

    @NotNull
    default CodeExecutor getCodeExecutor() {
      return new DefaultCodeExecutor();
    }

    @Nullable
    default TaskChecker getTaskChecker(@NotNull Task task, @NotNull Project project) {
        if (task instanceof EduTask) {
            return getEduTaskChecker((EduTask) task, project);
        }
        else if (task instanceof OutputTask) {
            return getOutputTaskChecker((OutputTask) task, project, getCodeExecutor());
        }
        else if (task instanceof TheoryTask) {
            return getTheoryTaskChecker((TheoryTask) task, project);
        }
        else if (task instanceof CodeTask) {
            return getCodeTaskChecker((CodeTask) task, project);
        }
        else if (task instanceof ChoiceTask) {
            return getChoiceTaskChecker((ChoiceTask) task, project);
        }
        else if (task instanceof IdeTask) {
            return getIdeTaskChecker((IdeTask) task, project);
        }
        else if (task instanceof CodeforcesTask) {
            if (task instanceof CodeforcesTaskWithFileIO) {
                return new CodeforcesTaskWithFileIOTaskChecker((CodeforcesTaskWithFileIO) task, project);
            }
            return new CodeforcesTaskChecker((CodeforcesTask) task, project, getCodeExecutor());
        }
        else {
            throw new IllegalStateException("Unknown task type: " + task.getItemType());
        }
    }
}

File: educational-core/src/com/jetbrains/edu/learning/checker/TaskCheckerProvider.kt
package com.jetbrains.edu.learning.checker

import com.intellij.openapi.project.Project
import com.jetbrains.edu.learning.codeforces.checker.CodeforcesTaskChecker
import com.jetbrains.edu.learning.codeforces.checker.CodeforcesTaskWithFileIOTaskChecker
import com.jetbrains.edu.learning.codeforces.courseFormat.CodeforcesTask
import com.jetbrains.edu.learning.codeforces.courseFormat.CodeforcesTaskWithFileIO
import com.jetbrains.edu.learning.courseFormat.tasks.*
import com.jetbrains.edu.learning.courseFormat.tasks.choice.ChoiceTask
import com.jetbrains.edu.learning.courseFormat.tasks.data.DataTask
import com.jetbrains.edu.learning.stepik.hyperskill.courseFormat.RemoteEduTask

/**
 * If you add any new methods here, please do not forget to add it also to
 * @see com.jetbrains.edu.learning.stepik.hyperskill.checker.HyperskillTaskCheckerProvider
 */
interface TaskCheckerProvider {
  val codeExecutor: CodeExecutor
    get() = DefaultCodeExecutor()

  val envChecker: EnvironmentChecker
    get() = EnvironmentChecker()

  fun getEduTaskChecker(task: EduTask, project: Project): TaskChecker<EduTask>

  fun getTheoryTaskChecker(task: TheoryTask, project: Project): TheoryTaskChecker = TheoryTaskChecker(task, project)

  // Should not be overloaded by anyone
  fun getTaskChecker(task: Task, project: Project): TaskChecker<*>? {
    return when (task) {
      is RemoteEduTask,
      is StringTask, is NumberTask, is CodeTask, is DataTask -> null

      is EduTask -> getEduTaskChecker(task, project)
      is OutputTask -> OutputTaskChecker(task, envChecker, project, codeExecutor)
      is TheoryTask -> getTheoryTaskChecker(task, project)
      is ChoiceTask -> if (task.canCheckLocally) ChoiceTaskChecker(task, project) else null
      is IdeTask -> IdeTaskChecker(task, project)
      is CodeforcesTaskWithFileIO -> CodeforcesTaskWithFileIOTaskChecker(task, project)
      is CodeforcesTask -> CodeforcesTaskChecker(task, envChecker, project, codeExecutor)
      else -> throw IllegalStateException("Unknown task type: " + task.itemType)
    }
  }
}